Addition of headless replay, overall refactor and addition of AcceptanceTestEnvironment

iaronaraujo opened the pull request:
- It is now possible to run ReplayTests in a headless environment, therefore they can be added to Jenkins.
- Overall refactor to improve the code.
- Creation of AcceptanceTestEnvironment. With it, it is now easier to create replay tests since it is only necessary to implement some abstract methods.

iaronaraujo committed:
It is now possible to run replay tests on a headless client
Updated Replay tests to work with the variable "position" rename to "index"

iaronaraujo committed:
Overall refactor and addition of AcceptanceTestEnvironment

iaronaraujo committed:
Added GameThread.reset() to the closing methods

Cervator commented on src/main/java/org/terasology/AcceptanceTestEnvironment.java:
```diff
@@ -0,0 +1,98 @@
+import org.terasology.recording.RecordAndReplayStatus;
+
+/**
+ * An environment responsible for setting up the workflow of a ReplayTest used in acceptance tests.
```
You could probably expand quite a bit on this javadoc, it is a pretty important class :-)

There are some great and detailed examples in some of the other `-Environment` classes

Cervator commented on src/main/java/org/terasology/AcceptanceTestEnvironment.java:
```diff
@@ -0,0 +1,98 @@
+    private boolean isHeadless;
+
+    /*
+     * To test the replay while it is executing, it is necessary to create a thread that will run the replay.
```
You could fit this on a one-liner javadoc instead 

 /** To test the replay while it is executing, it is necessary to create a thread that will run the replay. */

True, this wasn't javadoc in the first place, but it might as well be. It also is a little unusual since it isn't a simple field - it could be multi-line if there's enough doc to fill that :-)

Cervator commented on src/main/java/org/terasology/AcceptanceTestEnvironment.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    /**
+     * This method should have the @Test tag and it should call the "runTest" method.
```
Huh? It _does_ have the test tag! :-)

If there is such a tight contract and `runTest` is declared here as well would it make sense to simply declare the method fully right away?

Cervator commented on src/main/java/org/terasology/ReplayTestingEnvironment.java:
```diff
@@ -41,11 +47,13 @@
+import java.util.function.BooleanSupplier;
 
 /**
  * A base class for replay tests to inherit to run a replay in a thread while testing some variables.
```
Not part of the diff but this class could also use more expanded javadoc

Cervator commented on src/main/java/org/terasology/ReplayTestingEnvironment.java:
```diff
@@ -110,14 +115,18 @@ private void mainLoop() {
     /**
-     * Creates a headed TerasologyEngine.
-     * @return the created engine
+     * Creates a TerasologyEngine.
```
Another good spot for more javadoc

Cervator commented on src/test/java/org/terasology/replayTests/examples/templates/ReplayTestTemplate.java:
```diff
@@ -13,21 +13,22 @@
  * For more information about Replay Tests, see https://github.com/MovingBlocks/Terasology/wiki/Replay-Tests
  */
-public class ReplayTestTemplate extends ReplayTestingEnvironment { //Every replay test should extend ReplayTestingEnvironment
+public class ReplayTestTemplate extends ReplayTestingEnvironment { //Replay tests should extend ReplayTestingEnvironment
```
IMHO don't have a separate comment at the end here. Put it into the javadoc.

Cervator commented on src/test/java/org/terasology/replayTests/examples/templates/ReplayTestTemplate.java:
```diff
@@ -62,35 +68,22 @@ public void testTemplate() {
-                    //test something
-                }
-                */
+            //to test things in the middle of a replay, checks can be written between the 'waitUntils'
```
This is a great example of a bunch of comments / multiline comments that could all be moved into a beautifully formatted block inside the class level javadoc :-)

Cervator reviewed (approved):
Works great! With both https://github.com/MovingBlocks/Terasology/pull/3458 + https://github.com/MovingBlocks/Terasology/pull/3455 merged since both had different pieces needed here :-)

Actually a little puzzled that it works as the reset of the `GameThread` is happening from a module yet that class isn't on the whitelist. I guess security just isn't enabled in this context?

Tests work, although it is hard to tell from the logging. Not that a lot of tests really log things well, but maybe it would be nice to indicate via logging when something is being asserted? I dunno if it might get logged well enough if there's a failure. Have you tried running the tests negatively trying to make them break on purpose?

PR here works although the javadoc could be improved. Not letting that be a blocker if you want to do that in another PR but not merging this just yet in case of review comments from @eviltak or more javadoc etc :-)

iaronaraujo committed:
Javadoc improvements

iaronaraujo committed:
Updating git ignore

eviltak commented on src/main/java/org/terasology/ReplayTestingEnvironment.java:
```diff
@@ -168,6 +204,13 @@ protected boolean isInitialised() {
+    protected void waitUntil(BooleanSupplier supplier) {
+        try {
+            while (!supplier.getAsBoolean()) {
+                Thread.sleep(1000);
```
I still feel this should be changed to a smaller value, like `100` (10 FPS) or `50` (20 FPS). Ideally I'd wait for 16 ms (60 FPS). I think it is worth a try, and I don't see why sleeping for a shorter period of time should decrease performance.

eviltak commented on src/main/java/org/terasology/ReplayTestingEnvironment.java:
```diff
@@ -168,6 +204,13 @@ protected boolean isInitialised() {
+                Thread.sleep(1000);
+            }
+        } catch (Exception e) {
+            logger.error("Error has occurred in the waitUntil method: ", e);
```
Should we really be catching and logging exceptions here? An uncaught exception will fail the test and signify that there may be an unwanted problem in the code executed in `supplier.getAsBoolean()`. If the exception is caught and logged here, it might allow problems to slip under the radar. As a rule, I never catch exceptions in tests unless I _want_ that exception to occur (or want to manually fail the test via `Assert.fail`).

eviltak commented on src/main/java/org/terasology/AcceptanceTestEnvironment.java:
```diff
@@ -0,0 +1,98 @@
+                String replayTitle = recordingTitle;
+                AcceptanceTestEnvironment.super.openReplay(replayTitle, isHeadless);
+            } catch (Exception e) {
+                e.printStackTrace();
```
For reasons I state below, I believe exceptions should never be caught in a test (or a method used in a test). Also, won't this `catch` catch _all_ exceptions that occur during the execution of the replay test?

eviltak commented on src/main/java/org/terasology/AcceptanceTestEnvironment.java:
```diff
@@ -0,0 +1,98 @@
+            waitUntil(() -> getRecordAndReplayStatus() == RecordAndReplayStatus.REPLAY_FINISHED);
+            testOnReplayEnd();
+        } catch (Exception e) {
+            e.printStackTrace();
```
Same as above. The problem here is even more glaring since you catch all exceptions that may be thrown by the actual testing (`testXYZ`) methods, which is a big no-no.

eviltak commented on src/main/java/org/terasology/ReplayTestingEnvironment.java:
```diff
@@ -41,11 +47,27 @@
+ * {@link #isInitialised()} method.
+ * <p>
+ * The last protected method is {@link #waitUntil(BooleanSupplier)} which is generally used to wait for the replay to get
+ * to a certain point so something can be tested at the right moment.
```
I'd include class usage instructions rather than method descriptions here. The javadoc should describe the workflow for creating replay test by extending from `ReplayTestingEnvironment`. In the process of describing this workflow, important utility methods (like `getHost` and `waitUntil`) can be mentioned and briefly described (if necessary). The javadoc currently does not provide any information on how to create a replay test by extending this class, which methods to override, what the overriden method bodies should contain et cetera, which in my opinion is the more valuable information.

eviltak commented on src/main/java/org/terasology/AcceptanceTestEnvironment.java:
```diff
@@ -0,0 +1,98 @@
+ * <p>
+ * To write tests that uses this class, it is necessary to extend it and fill the three abstract methods {@link #testOnReplayStart()} ,
+ * {@link #testDuringReplay()} and {@link #testOnReplayEnd()} . If it is not desired to test something in one or more of
+ * the replay stages, it is possible to leave the methods in black.
```
I feel that this paragraph may be better utilized by describing when the three abstract methods will be called and what those method implementations could potentially contain. Right now, this paragraph contains redundant information which is either already enforced by the compiler/IDE ("all three methods need to be overriden"), or would be much better suited in the method specific documentation ("if it is desired, the methods may be blank").

eviltak commented on src/main/java/org/terasology/AcceptanceTestEnvironment.java:
```diff
@@ -0,0 +1,98 @@
+     * Tests that are executed when the replay ends.
+     * @throws Exception
+     */
+    protected abstract void testOnReplayEnd() throws Exception;
```
As stated in this class's javadoc review, these method descriptions should contain information which is currently included in the class javadoc, like a detailed description on when these methods are called and the fact that these can be left blank if desired.

eviltak commented on src/test/java/org/terasology/replayTests/WoodCutReplayTest.java:
```diff
@@ -33,32 +34,37 @@
-            WoodCutReplayTest.super.openReplay(replayTitle);
+            WoodCutReplayTest.super.openReplay(replayTitle, true);
         } catch (Exception e) {
             e.printStackTrace();
```
Same as above -- catching exceptions in a test.

eviltak commented on src/test/java/org/terasology/replayTests/WoodCutReplayTest.java:
```diff
@@ -33,32 +34,37 @@
     @Test
     public void testWoodcut() {
         replayThread.start();
         try {
```
Same as above -- catching exceptions in a test (can't comment on the catch block). Big no-no. If you've caught exceptions in tests elsewhere, I'd recommend you get rid of those `try...catch` blocks too.

eviltak commented on src/main/java/org/terasology/ReplayTestingEnvironment.java:
```diff
@@ -56,17 +78,18 @@
      */
-    protected void openReplay(String replayTitle) throws Exception {
-        host = createEngine();
+    protected void openReplay(String replayTitle, boolean isHeadless) throws Exception {
```
Should it also be pointed out that Jenkins (and/or other CI systems) may not be able to run headed replay tests and tests using headed replays should be `@Ignore`d?

eviltak commented on src/test/java/org/terasology/replayTests/examples/templates/ReplayTestTemplate.java:
```diff
@@ -0,0 +1,94 @@
+ * For more information about Replay Tests, see https://github.com/MovingBlocks/Terasology/wiki/Replay-Tests
+ * Replay tests should extend ReplayTestingEnvironment directly or indirectly.
+ */
+public class ReplayTestTemplate extends ReplayTestingEnvironment {
```
I'd move this to the wiki instead of leaving it in the code base. It'll be easier and cleaner to explain the usage of `ReplayTestingEnvironment` in the wiki using pretty markdown rather than having to explain it in javadoc.

iaronaraujo committed:
Refactor and Javadoc improvements.

Removed the ReplayTestTemplate class and ReplayTestingEnvironment isn't an abstract class anymore, making replay tests have one ReplayTestingEnvironment instead of extending it.
